File: api/Ussd.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Ussd 对应 Ussd.py。
type Ussd struct {
	*session.ApiGroup
}

// NewUssd 创建 Ussd API 分组。
func NewUssd(s *session.Session) *Ussd {
	return &Ussd{ApiGroup: session.NewApiGroup(s)}
}

// Status 对应 ussd/status。
func (u *Ussd) Status() (map[string]interface{}, error) {
	return u.S.Get("ussd/status", nil, "api")
}

// Get 对应 ussd/get。
func (u *Ussd) Get() (map[string]interface{}, error) {
	return u.S.Get("ussd/get", nil, "api")
}

// Send 发送 USSD 指令。对应 ussd/send (post_get)。
func (u *Ussd) Send(content string) (map[string]interface{}, error) {
	return u.S.PostGet("ussd/send", map[string]interface{}{
		"content":  content,
		"codeType": "codeType",
		"timeout":  nil,
	}, false, "api", false, false)
}

File: src/api/ussd.rs
//! Ussd API group (`api/Ussd.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_get_value};

/// Ussd API group.
pub struct Ussd<'a> {
    conn: &'a Connection,
}

impl<'a> Ussd<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Ussd { conn }
    }

    /// `ussd/status`. USSD status.
    pub fn status(&self) -> Result<Value> {
        get_value(self.conn, "api/ussd/status")
    }

    /// `ussd/get`. Current USSD session.
    pub fn get(&self) -> Result<Value> {
        get_value(self.conn, "api/ussd/get")
    }

    /// Send a USSD string.
    pub fn send(&self, content: &str) -> Result<Value> {
        let body: XmlMap = map_of([
            ("content", content.to_string()),
            ("codeType", "codeType".to_string()),
            ("timeout", "".to_string()),
        ]);
        post_get_value(self.conn, "api/ussd/send", &body)
    }
}
